### PR TITLE
[RFC] libvmaf/cuda: revisit vmaf_cuda_* api to support sharing of external CUcontext

### PR DESCRIPTION
**Pull request by kylophone**
This PR includes a change to the `vmaf_cuda_*` APIs to support an optional initialization with external `CUcontext`s. I have opted to modify the current `vmaf_cuda_*` API and have not targeted backwards compatibility. Since we have not had a libvmaf release with CUDA yet, this is not an API break. TL;DR for the API changes can be found in the API docs here: [include/libvmaf/libvmaf_cuda.h](https://github.com/Netflix/vmaf/compare/libvmaf_cuda_api?expand=1#diff-18ddbea75a7b00fbdd5ce78fd34b57f19dba40b0b2b49fb5dc4845fd860c727e)

When this new API is used from libavfilter and we import the FFmpeg-allocated `CUcontext`s, we can get about a 20% speedup (~300 FPS -> ~360 FPS). The FFmpeg patch for this new avfilter is [available here](https://github.com/Netflix/vmaf/files/11482368/vf_libvmaf_cuda.diff.zip). I will send this diff to ffmpeg-devel once we are certain the new API in this will work for us. Usage for `vf_libvmaf_cuda` is as follows:

```sh
./ffmpeg \
    -hwaccel cuda -hwaccel_output_format cuda -i dis.y4m.mp4 \
    -hwaccel cuda -hwaccel_output_format cuda -i ref.y4m.mp4 \
    -filter_complex libvmaf_cuda \
    -f null -
```

### CODE REVIEW

File: libvmaf/src/libvmaf.c [hunk 1]
```diff
@@ -42,13 +42,24 @@
...
+    struct {
+        unsigned w, h;
+        unsigned bpc;
+        enum VmafPixelFormat pix_fmt;
+    } pic_params;
+    enum VmafCudaPicturePreallocationMethod pic_prealloc_method;
+    int device_id;
+    int stream_priority;
+} VmafCudaConfiguration3;
+//^FIXME
```
Comment (gedoensmax) [inline libvmaf/src/libvmaf.c:62]: Due to the FIXME  i am guessing this 3 has to go away ?

**Review by gedoensmax — COMMENTED** [review]
Overall this looks good to me! The ffmpeg patch as well. 
I had to apply this patch on my side to remove some missing header occurences:
```
diff --git a/libvmaf/src/cuda/picture_cuda.c b/libvmaf/src/cuda/picture_cuda.c
index 96f84012..f615b9d8 100644
--- a/libvmaf/src/cuda/picture_cuda.c
+++ b/libvmaf/src/cuda/picture_cuda.c
@@ -21,7 +21,6 @@
 #include "mem.h"
 #include "picture_cuda.h"
 #include "common.h"
-#include "libvmaf/vmaf_cuda_state.h"
 #include "log.h"
 
 #include <cuda.h>
diff --git a/libvmaf/src/feature/cuda/integer_adm_cuda.c b/libvmaf/src/feature/cuda/integer_adm_cuda.c
index 043d261b..340ce421 100644
--- a/libvmaf/src/feature/cuda/integer_adm_cuda.c
+++ b/libvmaf/src/feature/cuda/integer_adm_cuda.c
@@ -22,7 +22,6 @@
 
 #include "cuda_helper.cuh"
 #include "mem.h"
-#include "libvmaf/vmaf_cuda_state.h"
 
 #include "feature_collector.h"
 #include "feature_extractor.h"
diff --git a/libvmaf/src/feature/cuda/integer_motion_cuda.c b/libvmaf/src/feature/cuda/integer_motion_cuda.c
index 137b780a..9615cde9 100644
--- a/libvmaf/src/feature/cuda/integer_motion_cuda.c
+++ b/libvmaf/src/feature/cuda/integer_motion_cuda.c
@@ -27,7 +27,6 @@
 #include "feature_extractor.h"
 #include "feature_name.h"
 #include "cuda/integer_motion_cuda.h"
-#include "libvmaf/vmaf_cuda_state.h"
 #include "mem.h"
 #include "picture.h"
 #include "picture_cuda.h"
diff --git a/libvmaf/src/feature/cuda/integer_vif_cuda.h b/libvmaf/src/feature/cuda/integer_vif_cuda.h
index dda3ac8b..dfb7165b 100644
--- a/libvmaf/src/feature/cuda/integer_vif_cuda.h
+++ b/libvmaf/src/feature/cuda/integer_vif_cuda.h
@@ -21,7 +21,6 @@
 #define FEATURE_VIF_CUDA_H_
 
 #include <stdint.h>
-#include "libvmaf/vmaf_cuda_state.h"
 #include "integer_vif.h"
 #include "common.h"
 
diff --git a/libvmaf/test/test_ring_buffer.c b/libvmaf/test/test_ring_buffer.c
index d0a1d864..abea74d5 100644
--- a/libvmaf/test/test_ring_buffer.c
+++ b/libvmaf/test/test_ring_buffer.c
@@ -25,7 +25,6 @@
 #include "test.h"
 
 #include "libvmaf/libvmaf_cuda.h"
-#include "libvmaf/vmaf_cuda_state.h"
 
 #include "cuda/common.h"
 #include "cuda/picture_cuda.h"
```

File: libvmaf/src/libvmaf.c (hunk 1 above)
Comment (kylophone) [reply libvmaf/src/libvmaf.c:62]: Yes, thanks. Meant to clean this up.

**Comment by kylophone** [comment]
> Overall this looks good to me! The ffmpeg patch as well. I had to apply this patch on my side to remove some missing header occurences:

Cool, thanks. I will add these changes to my patch.

**Comment by gedoensmax** [comment]
@kylophone have you thought about adding a GPU runner to the CI ? That would also enable correctness checks.

**Comment by jfroy** [comment]
I tried to apply this patch and the ffmpeg patch, but I get an error when I run the command:

```
libvmaf ERROR problem during cuda picture preallocation
    Last message repeated 2 times
[Parsed_libvmaf_cuda_0 @ 0x55fbe073b4c0] Failed to configure output pad on Parsed_libvmaf_cuda_0
Error reinitializing filters!
Failed to inject frame into filter network: Invalid argument
Conversion failed!
```

**Comment by kylophone** [comment]
> @kylophone have you thought about adding a GPU runner to the CI ? That would also enable correctness checks.

Yes, would be nice. Github actions doesn't have GPU instances, so we would need to figure out how to self-host.

**Comment by jfroy** [comment]
> > I tried to apply this patch and the ffmpeg patch, but I get an error when I run the command:
> > ```
> > libvmaf ERROR problem during cuda picture preallocation
> >     Last message repeated 2 times
> > [Parsed_libvmaf_cuda_0 @ 0x55fbe073b4c0] Failed to configure output pad on Parsed_libvmaf_cuda_0
> > Error reinitializing filters!
> > Failed to inject frame into filter network: Invalid argument
> > Conversion failed!
> > ```
> 
> Could you share the ffmpeg command you've used and uncut output?

```ffmpeg -v verbose \
                       -hwaccel cuda -hwaccel_output_format cuda -i input.mkv \
                       -hwaccel cuda -hwaccel_output_format cuda -i ref.mkv \
                       -filter_complex libvmaf_cuda \
                       -f null -
```

[ffmpeg-vmaf-log.txt](https://github.com/Netflix/vmaf/files/11558027/ffmpeg-vmaf-log.txt)

I built vmaf debug and the issue appears to be that `cfg.pix_fmt` is `VMAF_PIX_FMT_UNKNOWN` in `vmaf_cuda_preallocate_pictures`.

Both input streams are 10-bit 420 HEVC.

```
  Stream #0:0: Video: hevc (Main 10), 1 reference frame, yuv420p10le(tv, bt2020nc/bt2020/smpte2084, progressive, topleft), 3840x2160 (3840x2176) [SAR 1:1 DAR 16:9], 23.98 fps, 23.98 tbr, 1k tbn (default)
  Stream #1:0(eng): Video: hevc (Main 10), 1 reference frame, yuv420p10le(tv, bt2020nc/bt2020/smpte2084, topleft), 3840x2160 [SAR 1:1 DAR 16:9], 23.98 fps, 23.98 tbr, 1k tbn
```

**Comment by jfroy** [comment]
I built ffmpeg debug and now it's even more clear.

In `config_output_cuda`, `frames_ctx->sw_format` is `AV_PIX_FMT_P010LE`, which is not handled by `pix_fmt_map`, yielding `VMAF_PIX_FMT_UNKNOWN`.

**Comment by jfroy** [comment]
Adding `AV_PIX_FMT_P010LE` to `pix_fmt_map` seems to work and allows the computation to run.

**Comment by jfroy** [comment]
What sort of variance should I expect between the pure CPU pipeline and the pure GPU pipeline? Using this patch set, running with the same inputs, the CPU pipeline yields 90.617407 and the GPU pipeline yields 99.961091 (which is suspect).

CPU command:

```
ffmpeg \
                       -an -i dist.mkv \
                       -an -i ref.mkv \
                       -filter_complex libvmaf=model=version=vmaf_4k_v0.6.1:n_threads=8 \
                       -f null -
```

GPU command:

```
ffmpeg \
                       -hwaccel cuda -hwaccel_output_format cuda -an -i dist.mkv \
                       -hwaccel cuda -hwaccel_output_format cuda -an -i ref.mkv \
                       -filter_complex libvmaf_cuda=model=version=vmaf_4k_v0.6.1 \
                       -f null -
```

The inputs have the same resolution and frame rate (UHD 24 fps).

**Comment by kylophone** [comment]
> What sort of variance should I expect between the pure CPU pipeline and the pure GPU pipeline? 

Should be close to an exact match. If I had to guess, this could be a bug with the data copy from the new `AV_PIX_FMT_P010LE` pictures.

**Comment by kylophone** [comment]
> > > What sort of variance should I expect between the pure CPU pipeline and the pure GPU pipeline?
> > 
> > 
> > Should be close to an exact match. If I had to guess, this could be a bug with the data copy from the new `AV_PIX_FMT_P010LE` pictures.
> 
> That's quite possible. It's not because it's not erroring that it's computing correctly. I don't have a ton of time to look into this, but I'll trudge along.
> 
> I'm surprised no one else has run into this issue. If you're testing 10-bit stream, you're going to hit it. `ff_nvdec_frame_params()` (`src/ffmpeg/libavcodec/nvdec.c`) yields `AV_PIX_FMT_P010` for 10-bit non-444 streams.

I'll take a look at it. Thanks for testing the ffmpeg patch and finding this bug.

**Comment by jfroy** [comment]
> 

I've done some testing and it does not appear to change the result for my corpus. I'm collecting a few CC clips and testing with them to see if I can reproduce and I'll share that once ready.

**Comment by gedoensmax** [comment]
@jfroy could you generate detailed scores and share those when your testing is done ? Would be interesting to see from which feature extractor the changes are originating. For ffmpeg that would look something like `-filter_complex libvmaf_cuda=log_path=gpu.json:log_fmt=json`

**Comment by MorkTheOrk** [comment]
> This PR includes a change to the `vmaf_cuda_*` APIs to support an optional initialization with external `CUcontext`s. I have opted to modify the current `vmaf_cuda_*` API and have not targeted backwards compatibility. Since we have not had a libvmaf release with CUDA yet, this is not an API break. TL;DR for the API changes can be found in the API docs here: [include/libvmaf/libvmaf_cuda.h](https://github.com/Netflix/vmaf/compare/libvmaf_cuda_api?expand=1#diff-18ddbea75a7b00fbdd5ce78fd34b57f19dba40b0b2b49fb5dc4845fd860c727e)
> 
> When this new API is used from libavfilter and we import the FFmpeg-allocated `CUcontext`s, we can get about a 20% speedup (~300 FPS -> ~360 FPS). The FFmpeg patch for this new avfilter is [available here](https://github.com/Netflix/vmaf/files/11482368/vf_libvmaf_cuda.diff.zip). I will send this diff to ffmpeg-devel once we are certain the new API in this will work for us. Usage for `vf_libvmaf_cuda` is as follows:
> 
> ```shell
> ./ffmpeg \
>     -hwaccel cuda -hwaccel_output_format cuda -i dis.y4m.mp4 \
>     -hwaccel cuda -hwaccel_output_format cuda -i ref.y4m.mp4 \
>     -filter_complex libvmaf_cuda \
>     -f null -
> ```

@kylophone i saw that the FFMPEG patch contains cuMemcpy2D calls instead of cuMemcpy2DAsync, this is slower and should be updated.

**Pull request merged**